torch/distributions/inverse_wishart.py

```diff
@@ -0,0 +1,290 @@
+
+        new._batch_dims = [-(x + 1) for x in range(len(batch_shape))]
+
+        if 'covariance_matrix' in self.__dict__:
+            new.covariance_matrix = self.covariance_matrix.expand(cov_shape)
+        if 'scale_tril' in self.__dict__:
+            new.scale_tril = self.scale_tril.expand(cov_shape)
+        if 'precision_matrix' in self.__dict__:
+            new.precision_matrix = self.precision_matrix.expand(cov_shape)
+
+        # Chi2 distribution is needed for Bartlett decomposition sampling
+        new._dist_chi2 = torch.distributions.chi2.Chi2(
+            df=(
+                new.df.unsqueeze(-1)
+                - torch.arange(
+                    self.event_shape[-1],
+                    dtype=new._unbroadcasted_scale_tril.dtype,
+                    device=new._unbroadcasted_scale_tril.device,
+                ).expand(batch_shape + (-1,))
+            )
+        )
+
+        super(InverseWishart, new).__init__(batch_shape, self.event_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    @lazy_property
+    def scale_tril(self):
+        return self._unbroadcasted_scale_tril.expand(
+            self._batch_shape + self._event_shape)
+
+    @lazy_property
+    def covariance_matrix(self):
+        return (
+            self._unbroadcasted_scale_tril @ self._unbroadcasted_scale_tril.transpose(-2, -1)
+        ).expand(self._batch_shape + self._event_shape)
+
+    @lazy_property
+    def precision_matrix(self):
+        identity = torch.eye(
+            self._event_shape[-1],
+            device=self._unbroadcasted_scale_tril.device,
+            dtype=self._unbroadcasted_scale_tril.dtype,
+        )
+        return torch.cholesky_solve(
+            identity, self._unbroadcasted_scale_tril
+        ).expand(self._batch_shape + self._event_shape)
+
+    @property
+    def mean(self):
+        p = self._event_shape[-1]  # has singleton shape
+        if self.df.le(p + 1).any():
+            raise ValueError("Mean of the Inverse Wishart distribution can be caculated only for df > ndim + 1.")
+        return self.covariance_matrix.div(self.df - 1)
+
+    @property
+    def variance(self):
+        nu = self.df  # has shape (batch_shape)
+        p = self._event_shape[-1]  # has singleton shape
+        V = self.covariance_matrix  # has shape (batch_shape x event_shape)
+        diag_V = V.diagonal(dim1=-2, dim2=-1)
+        eff_df = (nu - p).view(self._batch_shape + (1, 1))
+
+        return (
+            (eff_df + 1) * V.pow(2)
+            + (eff_df - 1) * torch.einsum("...i,...j->...ij", diag_V, diag_V)
+        ) / (eff_df * (eff_df - 1).pow(2) * (eff_df - 3))
+
+    def _bartlett_sampling(self, sample_shape=torch.Size()):
+        p = self._event_shape[-1]  # has singleton shape
+
+        # Implemented Sampling using Bartlett decomposition
+        noise = self._dist_chi2.rsample(sample_shape).sqrt().reciprocal().diag_embed(dim1=-2, dim2=-1)
```

Comment:
Could you mention what are the changes w.r.t. the Wishart distribution implementation? e.g. reciprocal of Chi2 samples here. Any other? There's a lot of duplicated code between these two classes and we should think about how to restructure to keep redundancy to a minimum. If there are many small subtle variations, then this isn't important and we can keep them separate.

Reply:
Thank you for discovering good point to improve.
I seperated the implementation of Wishart and Inverse Wishart distribution before because of 2 reasons:
1) Some method or property like `.mean()` can be calculated in different conditions of Wishart and Inverse Wishart. 
2) I expected there is more sophisticated sampling algorithm for Inverse Wishart which cannot applied in Wishart.
But I think I have to admit that two implementation shares too much duplicated parts.

Those are my references about the latest sampling techniques for Inverse Wishart:
[1] [Student-t Processes as Alternatives to Gaussian Processes](https://arxiv.org/abs/1402.4306)
[2] [Scalable Bayesian dynamic covariance modeling with variational Wishart and inverse Wishart processes](https://arxiv.org/abs/1906.09360)
[3] [A variational approximate posterior for the deep Wishart process](https://arxiv.org/abs/2107.10125)
I hope that me and other users in the future to easily apply Variational Inference techniques for Matrix like in [3] using the official implementation in PyTorch.
I will be happy to restructure and find a way to unite two class in one implementation while fitting this requirements.